pil_analyzer/src/evaluator.rs

```diff
@@ -40,65 +44,253 @@ pub fn compute_constants<T: FieldElement>(
         .collect()
 }
 
+#[derive(Clone)]
+pub enum Value<T> {
+    Number(T),
+    String(String),
+    Tuple(Vec<Value<T>>),
+    Array(Vec<Value<T>>),
+    // TODO check that the environment and the local variable numbering works.
+    // numbering is done in pil_analyzer: process_function
+    Closure(Expression<T>, Vec<Value<T>>),
+}
+
+impl<T: Eq> PartialEq for Value<T> {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Number(l), Self::Number(r)) => l == r,
+            (Self::String(l), Self::String(r)) => l == r,
+            (Self::Tuple(l), Self::Tuple(r)) => l == r,
+            (Self::Array(l), Self::Array(r)) => l == r,
+            (Self::Closure(_, _), Self::Closure(_, _)) => panic!("Tried to compare closures."),
```

Comment:
It seems like introducing a `Closure` type and putting the panic in its `PartialEq` impl would allow to derive `PartialEq` for `Value`, reducing boilerplate

Reply:
True! Also Closure is a rather complicated type.